### PR TITLE
Added event listener to remove unsubscribe websocket if user removed …

### LINKED ISSUE #884: Support using WebSocket subscriptions with onchain privacy groups with multi-tenancy
WS subscriptions can only be created for a privacy group you are currently a member of and the filter will be removed when you're removed from a group.

Need to confirm and document behaviour if you're removed on a reorg and the filter is removed but you end up still a member of the group.

### PR DESCRIPTION
**Pull request by mark-terry**
…from privacy group.

Signed-off-by: Mark Terry <mark.terry@consensys.net>

## PR description
Updated the multi-tenancy websocket management to unsubscribe if a subscribed user is removed from the corresponding privacy group.

## Fixed Issue(s)

Fixes #884 

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/SubscriptionManager.java [hunk 1]
```diff
@@ -158,4 +161,20 @@ public void sendMessage(final Long subscriptionId, final JsonRpcResult msg) {
...
+    subscriptionsOfType(SubscriptionType.LOGS, PrivateLogsSubscription.class).stream()
+        .filter(
+            subscription ->
+                subscription.getEnclavePublicKey().equals(event.getEnclavePublicKey())
+                    && subscription.getPrivacyGroupId().equals(event.getPrivacyGroupId()))
+        .forEach(
+            subscription ->
+                this.unsubscribe(
+                    new UnsubscribeRequest(
+                        subscription.getSubscriptionId(), subscription.getConnectionId())));
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/SubscriptionManager.java:178]: do we have to worry about this request failing?

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/PrivSubscribeTest.java [hunk 2]
```diff
@@ -134,6 +137,31 @@ public void multiTenancyCheckFailure() {
         .hasMessageContaining("msg");
   }
 
+  @Test
+  public void multiTenancyCheckSuccess() {
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/PrivSubscribeTest.java:141]: is there a test to check multiTenancyCheckFail?

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/request/SubscriptionRequestMapperTest.java [hunk 3]
```diff
@@ -409,7 +413,8 @@ public void mapRequestToPrivateSubscriptionWithInvalidType() {
     thrown.expect(InvalidSubscriptionRequestException.class);
     thrown.expectMessage("Invalid subscribe request. Invalid private subscription type.");
 
-    mapper.mapPrivateSubscribeRequest(new JsonRpcRequestContext(jsonRpcRequest));
+    mapper.mapPrivateSubscribeRequest(
+        new JsonRpcRequestContext(jsonRpcRequest), "enclave_public_key");
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/request/SubscriptionRequestMapperTest.java:417]: should this enclavePublicKey be the variable not the string

**Review by macfarla — COMMENTED** [review]
Looks good overall - couple of comments

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/SubscriptionManager.java [hunk 4]
```diff
@@ -158,4 +161,20 @@ public void sendMessage(final Long subscriptionId, final JsonRpcResult msg) {
           }
         });
   }
+
+  @Override
+  public void onPrivateTransactionProcessed(final PrivateTransactionEvent event) {
+    // When a user is removed from a privacy group, remove all subscriptions from that user to that
+    // group
+    subscriptionsOfType(SubscriptionType.LOGS, PrivateLogsSubscription.class).stream()
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/SubscriptionManager.java:169]: Do you need to handle the same race condition as filters (eg user removed from group but then in the same block requests to create a new subscription) - which we solved by storing the private tx events until the block was processed.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/PrivSubscribeTest.java (hunk 2 above)
Comment (mark-terry) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/PrivSubscribeTest.java:141]: As far as I can tell, the previous `multiTenancyCheckFailure` test checks the multi-tenancy fail case.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/SubscriptionManager.java (hunk 1 above)
Comment (mark-terry) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/SubscriptionManager.java:178]: I don't believe so. The fail case is if the specified connection doesn't exist, which shouldn't occur given we're retrieving those details from the current subscription map.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/SubscriptionManager.java (hunk 4 above)
Comment (mark-terry) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/SubscriptionManager.java:186]: PR updated to address this.

**Review by macfarla — APPROVED** [review]
LGTM

**Pull request merged**